src/cmap/conn/wire/message.rs

```diff
@@ -75,33 +75,38 @@ impl Message {
     }
 
     /// Reads bytes from `reader` and deserializes them into a Message.
-    pub(crate) async fn read_from(reader: &mut AsyncStream) -> Result<Self> {
-        let header = Header::read_from(reader).await?;
-        let mut length_remaining = header.length - Header::LENGTH as i32;
+    pub(crate) async fn read_from(
+        reader: &mut AsyncStream,
+        partial_message_state: &mut PartialMessageState,
+    ) -> Result<Self> {
+        let header = Header::read_from(reader, partial_message_state).await?;
 
         let flags = MessageFlags::from_bits_truncate(reader.read_u32().await?);
-        length_remaining -= std::mem::size_of::<u32>() as i32;
+        partial_message_state.bytes_remaining -= std::mem::size_of::<u32>();
 
         let mut count_reader = CountReader::new(reader);
         let mut sections = Vec::new();
 
-        while length_remaining - count_reader.bytes_read() as i32 > 4 {
+        while partial_message_state.bytes_remaining - count_reader.bytes_read() > 4 {
             sections.push(MessageSection::read(&mut count_reader).await?);
```

Comment:
the state needs to be propagated down here too. I wonder if this would be simpler if we just spawned a task via `RUNTIME.spawn` in `execute_operation` to guarantee everything got polled to completion. Doing all this partial reading and passing of state around seems to be pretty error-prone.

Reply:
Can you elaborate on what you mean by spawning a task in execute operation? Do you mean running `send_command` in a background task and then using a channel to get the result back?

Reply:
yeah, except a channel wouldn't be needed. We could just use our [`AsyncJoinHandle`](https://github.com/mongodb/mongo-rust-driver/blob/master/src/runtime/join_handle.rs#L15).

Reply:
Thinking about it more, it would probably be safest to do the entirety of `execute_operation`'s body in a task, so that the various things it needs to do based on operation execution happen (like updating the cluster time, clearing the pool on certain errors, etc.)

Reply:
I just tried to play around with this; it runs into some issues due to the fact that we can only spawn futures that are 'static, which won't be the case if we need to capture over any references. When trying to work around this, I ran into issues with the cursor's getMore operation, so I tried rebasing your sessions PR to see if the new cursor implementation avoided this issue, but this actually made things even more difficult due to the need to capture a reference to the session, which doesn't have the static lifetime.